# 4.3 Edit distance, assembly, overlaps

* Finish the read aignment problem.
* Dynamic programming algorithms for edit distance.
* Dynamic programming can be used for approximate matching
* The genome assembly problem

## The edit distance problem
* Another way to solve the approximate matching problem
* Can also be used to solve the alignment problem
* Hamming distance and edit distance are ways of measuring the difference between strings

**Hamming distance** is the number of number of mismatched bases between two sequences of the same length.

**Edit distance** is the minimum difference between two sequences of any length (includes insertions and deletions).

#### Edit distance
if |X| = |Y| 
editdistance(X, Y) ≤ hammingdistance(X, Y).

if |X| ≠ |Y|
editdistance(X, Y) ≥ ||X| - |Y||

For an edit distance algorithm is is useful to compare prefixes.

For two sequences:

    α x 
    β y

edit distance is the minimum of three things:
* The diffence between α and β (+1 if x ≠ y) (hamming distance)
* The difference between αx and β +1 (deletion in β)
* The difference between α and βy +1 (deletion in α)

```
edist(αx, βy) = min(
    edist(α, β) + δ(x, y),
    edist(αx, β) + 1,
    edist(α, βy) + 1
)
δ(x, y) = 0 if x = y else 1
```

This can be implemented as a recursive function.

For example: 

```python
def edDistRecursive(a, b):
    if len(a) == 0:
        return len(b)
    if len(b) == 0:
        return len(a)
    delt = 1 if a[-1] != b[-1] else 0
    return min(
        edDistRecursive(a[:-1], b[:-1]) + delt,
        edDistRecursive(a[:-1], b) + 1,
        edDistRecursive(a, b[:-1]) + 1
    )
```
This function works, but it is very very slow (O e^n).

### Dynamic programming for edit distance
* Recursive functions are slow because for each function call, the function is called 3 more times, etc.
* Inevitably, the function is called with the same arguments multiple times. 
* We will rewrite the function using a matrix which stores the edit distance for each prefix combination. 
* Each cell of the matrix corresponds to a pair of arguments which call the function only once.
* We will use the same expression for calculating the edit distance.
* This algorithm is much faster.

### Practical
[Edit distance](https://github.com/BenLangmead/ads1-notebooks/blob/master/3.01_EditDistanceDP.ipynb)

## New approximate matching solution
* How can we apply this edit distance algorithm to the approximate matching problem?
* A matrix is made using the text, T and pattern, P.
* Instead of initialising the first row with the prefix length of T, we will just fill with zeroes. 
* This is because we don't know where P will occur in T; every offset is equally likely, so all are 0.
* Once the matrix is filled in with edit distances, we need to identify approximate matches.
* By looking in the last row (full pattern), we can find the smallest edit distance. We don't yet know what the substring of T is.
* We can find out how we got here by looking at the neighnouring cells to find the smallest. This process is repeated until we reach the top row. This path is called the traceback.
* The traceback gices us the offset and substring of P. It also tell us the 'shape' of the approximate match, ie the locations of insertions and deletions.
* The problem with this algorithm is that the efficiency is related to the size of the matrix (|P| * |T|) which can get too big.
* In practice, edit distance-based approximate matching is used with other methods (pigeonhole, indexing, etc)to improve efficiency.

## Global and local alignment
* Global alignment is the end to end alignment of two strings.
* Local alignment is the alignment of multiple substrings of two strings. This can be used to find rearrangements.
* Semi-global alignment is the alignment of single substrings of two strings. This is useful to account for 


* Using dynamic programming to solve global and local alignment problems.
* Edit distance penalises all mismatch types equally, but this doesn't always make sense. If some substitutions are more common, they should be penalised less.
* DNA substitutions can be divided into transitions (purine -> purine) or transversions (purine -> pyrimidine). 
    * There are twice as many possibilities for transversions than transitions, but transitions are actually twice as frequent. 
    * Transversions should be penalised more.
* Indels (insertions and deletions) are mush less common than substitutions (maybe because frame shift?)
* We can design a penalty matrix as follows.


    2: transitions
    4: transversions
    8: gaps

| |A|C|G|T|-|
|-|-|-|-|-|-|
|A|0|4|2|4|8|
|C|4|0|4|2|8|
|G|2|4|0|4|8|
|T|4|2|4|0|8|
|-|8|8|8|8|

* The values of this matrix can be changed depending on the requrements of the sequences being compared.
* All we need to change in our algorithm is the inclusion of penalties as follows:

```
galgn(αx, βy) = min(
    galgn(α, β) + p(x, y),
    galgn(αx, β) + p(x, -),
    galgn(α, βy) + p(-, y)
)
```

**Global alignment** sets different penalties based on the biological problem.

![Global alignment](https://bio.libretexts.org/@api/deki/files/28393/page66image65266832.png?revision=1)

**Local alignment** finds the most similar substrings. 

![Local alignment](https://bio.libretexts.org/@api/deki/files/28370/page66image65266624.png?revision=1)

* This is difficult because we have to compare every possible pair of substrings (O n<sup>2</sup>).
* But using the dynamic programming algorithm these comparisons aren't actually necessary.
* We use a scoring matrix which gives a positive score for a match and a negative score for a mismatch.
* When the matrix is created, most values will be 0 (no similarity). We want to find the closest matches with higher scores.
* To find the traceback we look for the highest scores and follow them back to a 0.

### Practical
[Global alignment](https://github.com/BenLangmead/ads1-notebooks/blob/master/3.02_GlobalAlignment.ipynb)